bwl: nl80211_client: add set_tx_power_limit API

tomereli opened the pull request:
Some EasyMesh flows require setting tx power limit, which is not
supported via hostapd (tx_power is readable via radio_info command),
Add set_tx_power_limit() API to the nl80211 client implementation which
uses NL80211_CMD_SET_WIPHY to set the limit.

Currently doesn't work since set tx power is not supported by the driver, but this can be merged since we get the same status as when doing this with `iw` - so marking as ready for review.
**Tested on AXEPOINT3000 (UGW) and everything works as expected.**

Since this PR changes some channel selection handling, triggered gitlan CI pipeline to verify nothing broke, which tests the following tests and verifies autoconfig, onboarding and channel selection:
```
MAP-4.2.1
MAP-4.5.2_ETH_FH5GL
MAP-4.5.3_ETH_FH5GL
MAP-4.5.3_ETH_FH5GH
```
https://gitlab.com/prpl-foundation/prplMesh/pipelines/128634565

Fixes #983
 
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

mariomaz commented on common/beerocks/bwl/dummy/nl80211_client_dummy.h:
```diff
@@ -44,6 +44,18 @@ class nl80211_client_dummy : public nl80211_client {
+     *
+     * Set tx power limit for a radio
+     *
+     * @param[in] local_interface_name Virtual AP (VAP) interface name.
```
I think it can be a physical radio interface as well as a virtual interface

tomereli replied:
I think it will cause confusion, so lets keep it for radio only.

mariomaz commented on common/beerocks/bwl/include/bwl/ap_wlan_hal.h:
```diff
@@ -325,7 +325,10 @@ class ap_wlan_hal : public virtual base_wlan_hal {
      * @return true on success or false on error.
      */
-    virtual bool set_tx_power_limit(int tx_pow_limit) = 0;
+    virtual bool set_tx_power_limit(int tx_pow_limit)
```
I think parameter should be uint32_t

tomereli replied:
This was already like that apparently and I reverted this original change so I prefer not to change this as part of this PR.

mariomaz commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -72,6 +72,15 @@ class nl80211_client {
+     * @brief Set the tx power limit object
+     *
+     * Set tx power limit for a radio
+     * @param[in] local_interface_name Virtual AP (VAP) interface name.
```
I think it works both with radio or virtual interface name

tomereli replied:
I checked on UGW where I have only radio and it works, I didn't check VAP so I'll change the comment to say only the radio name is supported.

mariomaz commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -182,4 +182,34 @@ bool nl80211_client_impl::get_sta_info(const std::string &local_interface_name,
+            return true;
+        },
+        [&](struct nl_msg *msg) -> bool {
+            return true;
```
There must be some kind of return code to check if operation succeeded or not

tomereli replied:
I couldn't find this - from what I could understand, there is a default callback for error so I don't need to handle errors here.

mariomaz commented on common/beerocks/bwl/shared/nl80211_client_impl.h:
```diff
@@ -57,6 +57,17 @@ class nl80211_client_impl : public nl80211_client {
+     * @brief Set the tx power limit object
+     *
+     * Set tx power limit for a radio
+     * @param[in] local_interface_name Virtual AP (VAP) interface name.
```
Radio interface

mariomaz commented on common/beerocks/bwl/shared/nl80211_client_impl.h:
```diff
@@ -57,6 +57,17 @@ class nl80211_client_impl : public nl80211_client {
                               const sMacAddr &sta_mac_address, sStaInfo &sta_info) override;
 
+    /**
+     * @brief Set the tx power limit object
```
What do you mean by 'object'?

tomereli replied:
typo

mariomaz commented on common/beerocks/bwl/common/base_wlan_hal.cpp:
```diff
@@ -32,6 +33,9 @@ base_wlan_hal::base_wlan_hal(HALType type, std::string iface_name, IfaceType ifa
     // Initialize complex containers of the radio_info structure
     m_radio_info.supported_channels.resize(128 /* TODO: Get real value */);
+
+    m_nl80211_client = nl80211_client_factory::create_instance();
```
nl80211_client is a dependency and thus should be injected, not hard-coded.
If we ever create a unit test for this class, nl80211_client should be provided as a parameter.

tomereli replied:
I'm not sure I follow - can you explain what is hard coded here and what do you mean by injecting it?

mariomaz replied:
`nl80211_client` is a dependency for `base_wlan_hal`. `base_wlan_hal` should not create it. `nl80211_client` should be injected as a parameter to the constructor of `base_wlan_hal`.

But this is only actually required if we want to create a unit test for `base_wlan_hal`. In such unit test, we would pass a mock `nl80211_client` instead of the real implementation.

tomereli replied:
Done

mariomaz reviewed (approved):
Only nitpick comments so approving.

mariomaz commented:
BTW, I've realized just now that you've moved `set_tx_power_limit` from `ap_wlan_hal_*` to `base_wlan_hal`. This is a base class for `ap_wlan_hal_*`, `base_wlan_hal_*`, `mon_wlan_hal_*` and `sta_wlan_hal_*`classes so `set_tx_power_limit` will be available to all of them. Was that the intention or a side effect? I think there will be more instances of `nl80211_client` that what they are really needed.

Moreover, since `ap_wlan_hal_*` classes extend both `base_wlan_hal` and `base_wlan_hal_*` there will be two instances of `nl80211_client`. Again, more than needed. 

I think `nl80211_client` should be a member of `ap_wlan_hal_*` and not of `base_wlan_hal`

tomereli commented:
> BTW, I've realized just now that you've moved `set_tx_power_limit` from `ap_wlan_hal_*` to `base_wlan_hal`. This is a base class for `ap_wlan_hal_*`, `base_wlan_hal_*`, `mon_wlan_hal_*` and `sta_wlan_hal_*`classes so `set_tx_power_limit` will be available to all of them. Was that the intention or a side effect? I think there will be more instances of `nl80211_client` that what they are really needed.
> 
> Moreover, since `ap_wlan_hal_*` classes extend both `base_wlan_hal` and `base_wlan_hal_*` there will be two instances of `nl80211_client`. Again, more than needed.
> 
> I think `nl80211_client` should be a member of `ap_wlan_hal_*` and not of `base_wlan_hal`

True, didn't think about this. Thanks. I added the don't merge tag for now.

tomereli commented:
@mariomaz I refactored this PR based on your last comment, so I think this requires a new review.

arnout commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -72,6 +72,16 @@ class nl80211_client {
     virtual bool get_sta_info(const std::string &local_interface_name,
                               const sMacAddr &sta_mac_address, sStaInfo &sta_info) = 0;
+    /**
+     * @brief Set the tx power limit object
```
Didn't you say that "object" was a typo and that you fixed it?

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -713,9 +715,11 @@ update_vap_credentials_configure_wpa(const std::string &vap_if,
     : base_wlan_hal(bwl::HALType::AccessPoint, iface_name, IfaceType::Intel, callback, hal_conf),
-      base_wlan_hal_dwpal(bwl::HALType::AccessPoint, iface_name, callback, hal_conf)
+      base_wlan_hal_dwpal(bwl::HALType::AccessPoint, iface_name, callback, hal_conf),
+      m_nl80211_client(std::move(nl80211_client))
```
I believe the std::move is redundant here because unique_ptr only has a move constructor, no copy constructor. (nitpick)

tomereli replied:
Apparently it is not :/

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -2521,7 +2524,10 @@ bool ap_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
+    LOG_IF(!nl80211_client, FATAL) << "Failed to create nl80211_client instance";
+    return std::make_shared<dwpal::ap_wlan_hal_dwpal>(iface_name, callback, hal_conf,
+                                                      std::move(nl80211_client));
 }
```
@mariomaz I still don't really understand how the mocking would work. Now the nl80211_client_factory is hardcoded here, in ap_wlan_hal_create. If we would want to create a unit test for ap_wlan_hal_dwpal, we'd have to create a "mocking" factory function to replace this one, correct?

tomereli replied:
I moved the `create_instance` to `base_wlan_hal_XXX` constructor since the nl80211_client can be and will be used by the base class to update radio info which is implemented in the base class.

As @vitalybu said:
>Dependency injection is useful when you want to "inject" different implementations of the same interface to some class. I'm just not sure if it's the case here. Why would we ever use a non-DWPAL nl80211 client with DWPAL BWL?

arnout commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -327,7 +327,8 @@ bool ap_wlan_hal_dummy::read_supported_channels() { return true; }
 {
-    LOG(TRACE) << __func__ << " power limit: " << tx_pow_limit;
+    LOG(TRACE) << __func__ << " setting power limit: " << tx_pow_limit * 100 << " mBm";
+    m_radio_info.conducted_power = tx_pow_limit * 100;
```
IMO this could have been squashed into the first commit.

tomereli replied:
IMHO it makes more sense as a separate commit.

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -582,8 +582,24 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+                    LOG(ERROR) << "Failed building message!";
+                    return false;
+                }
+                ap_wlan_hal->refresh_radio_info();
```
Since nothing has changed, refreshing radio info shouldn't be needed.

Looking at it in more detail, it's even worse. Setting the TX power will not be reflected in the result of get_tx_power(), because that is using the tx power limit from the supported channels, and that, in turn, is purely based on the regulatory limits and the hardware limit, *not* on the power limit set by NL80211_ATTR_WIPHY_TX_POWER_SETTING. So I think we need to add an actual power limit setting in bwl::RadioInfo and make sure it gets filled in by refresh_radio_info(). But that's for another PR...

tomereli replied:
It looks like @mariomaz already included the max tx power using `NL80211_FREQUENCY_ATTR_MAX_TX_POWER` in #977 - I think we should decide how we want to proceed with @mariomaz 's PR before we merge this one.
I have UGW so I can test it correctly all the way, no need to rush and merge this half way done.

tomereli replied:
So apparently my previous comment is wrong, we don't need to rebase on #977 in order to read the tx power since it is reflected in `TxPower` when running `hostapd_cli -iwlan0 radio_info` hence available in `m_radio_info.conducted_power`. Added a commit which fixes this.

arnout reviewed (approved):
Just one remaining typo in a comment.

Commit:
bwl: nl80211_client: add set_tx_power_limit API

Some EasyMesh flows require setting tx power limit, which is not
supported via hostapd (tx_power is readable via radio_info command),
Add set_tx_power_limit() API to the nl80211 client implementation which
uses NL80211_CMD_SET_WIPHY to set the limit.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl:base_wlan_hal_dwpal: add nl80211_client member

Preparative commit.
The nl80211_client provides APIs to get and set WLAN parameters via
NL80211. It will be used by the AP manager to set the tx power limit,
and by the base_wlan_hal_dwpal to read radio info parameters which are
not available from hostapd, from the nl80211 client using
NL80211_CMD_GET_WIPHY.

This commit adds the nl80211_client member and set_tx_power_limit
implementation by calling the nl80211_client set_tx_power_limit API.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl:ap_wlan_hal_nl80211: add nl80211_client member

The nl80211_client provides APIs to get and set WLAN parameters via
NL80211. It will be used by the AP manager to set the tx power limit,
and by the base_wlan_hal_80211 to read radio information which is not
available from hostapd, from the nl80211_client via
NL80211_CMD_GET_WIPHY command.

This commit adds the nl80211_client member and set_tx_power_limit
implementation by calling the nl80211_client set_tx_power_limit API.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: ap_wlan_hal_dummy: add set_tx_power_limit implementation

Preparative commit.
We simply update the conducted power which is read in
refresh_radio_info and will be used by test_flows to verify the tx_power
limit setting as part of channel selection test.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: ap manager: add fill_cs_params helper func

Preparative commit.
CS (channel selection) parameters are sent from the ap manager once on
init in the join message, and then whenever there is a CSA or ACS event.

In case the AP manager is requested to set a new tx power limit, it will
need to send a CSA notification to the son slave so this in turn can
send the operating channel report in case only the tx power has been
changed.

As a preparation, add fill_cs_params private function to the ap manager
which is currently used only by the hal event handler but later will
also be used when only the tx power is changed.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ap manager: send CSA notification on txpower change

When the ACTION_APMANAGER_HOSTAP_CHANNEL_SWITCH_ACS_START requires a
change only for tx power, there will be no CSA event which will trigger
the sending of the operating channel report by the agent since there is
no event from hostapd when only the tx power is changed.

Therefore, simulate CSA event in the ap manager when only the tx limit
change is requested and send a CSA notification to the agent.

Refresh radio info after calling the
set_tx_power_limit API ensures that the new tx power read from hostapd
via the radio_info command will reflect the new limit settings.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: remove power limit workaround

Up till now, there was no event from the ap manager when only the tx
power was changed, and since this is required by the specification, we
manually sent the operating channel report.
The previous commit added a notification when the tx power is changed,
so this workaround is not needed anymore.
Send the operating channel report when hdnling the channel selection
request only if both tx power and channel switch are both not required.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: refresh radio info on CSA/ACS event

Refresh radio info on ACS_Completed/CSA_Finished so we report the latest
channel switch parameters in the notification.

This solves a potential bug where channel selection response will use
old channel switch parameters.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: use tx_power from hostapd RadioInfo TxPower

The current `get_tx_power` is wrong since it reads the regulatory domain
and not based on actual tx power, as explained in
https://github.com/prplfoundation/prplMesh/pull/982#discussion_r395548381.

The current Tx Power is reported in the TxPower parameter when running
hostapd_cli -iwlanX radio_info, therefore use it instead.
This was tested on AXEPOINT3000 with iw command to set tx power and then
hostapd_cli radio_info to verify this indeed reflects the limit.

Also delete the now unused get_tx_power function.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -182,4 +182,32 @@ bool nl80211_client_impl::get_sta_info(const std::string &local_interface_name,
+    // Get the interface index for given interface name
+    int iface_index = if_nametoindex(local_interface_name.c_str());
+    if (0 == iface_index) {
+        LOG(ERROR) << "Failed to read the index of interface " << local_interface_name << ": "
```
I think that you should either validate that `local_interface_name` is not empty, or at least add `''` to the log (so that it'll be easier to spot an empty value):
```suggestion
        LOG(ERROR) << "Failed to read the index of interface '" << local_interface_name << "': "
```

vitalybu commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -182,4 +182,32 @@ bool nl80211_client_impl::get_sta_info(const std::string &local_interface_name,
+        NL80211_CMD_SET_WIPHY, 0,
+        [&](struct nl_msg *msg) -> bool {
+            nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+            nla_put_u32(msg, NL80211_ATTR_WIPHY_TX_POWER_SETTING, NL80211_TX_POWER_LIMITED);
```
Shouldn't we allow returning the power setting to `NL80211_TX_POWER_AUTOMATIC`?
Or perhaps even support all TX_POWER modes:
- NL80211_TX_POWER_AUTOMATIC
- NL80211_TX_POWER_FIXED
- NL80211_TX_POWER_LIMITED

tomereli replied:
The only thing which is needed by EasyMesh is setting power limit.

vitalybu replied:
¯\\_(ツ)_/¯

vitalybu commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1590,8 +1594,7 @@ bool ap_wlan_hal_dwpal::read_supported_channels()
 {
-    LOG(DEBUG) << "set_tx_power_limit(): missing function implementation";
-    return true;
+    return m_nl80211_client->set_tx_power_limit(m_radio_info.iface_name, tx_pow_limit);
```
Since now `nl80211_client` is passed to the constructor (and has a default value of `nullptr`), you can end up with a null reference exception here.
You should either remove the default value and make sure it's not `nullptr`, or check if it's not `nullptr` before every call.

tomereli replied:
But in dwpal we always pass it, and override the default value - and if it fails then we assert with LOG_IF(FATAL) - do you still think it is necessary to check every call?

vitalybu replied:
No, I think you should remove the default value from the constructor and force passing a valid instance, while also validating that it's not nullptr.

tomereli replied:
OK, done. I removed the default value and just moved the initialization to `base_wlan_hal_*`. IMO it makes much more sense there instead of manually specifying it for every derived class.

vitalybu commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -572,8 +576,7 @@ bool ap_wlan_hal_nl80211::read_supported_channels()
 {
-    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
-    return true;
+    return m_nl80211_client->set_tx_power_limit(m_radio_info.iface_name, tx_pow_limit);
```
dwpal (previous commit) comments apply here as well

Commit:
tree-wide: rename "conducted_power" to "tx_power"

The mixed terminology causes confusion, so use tx_power all over instead
of conducted_power, which also aligns to the hostapd terminology.

This is a mechanical change using editor search-replace feature.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h:
```diff
@@ -88,6 +89,8 @@ class base_wlan_hal_dwpal : public virtual base_wlan_hal,
     bool dwpal_nl_cmd_scan_dump();
     void *get_dwpal_nl_ctx() const { return (m_dwpal_nl_ctx); }
 
+    std::unique_ptr<nl80211_client> m_nl80211_client;
```
What is the point of moving it to base_wlan_hal_dwpal? I though we said that it wouldn't be used by the monitor or station HALs...

More importantly though: if you move it here, it should also be taken as an argument in the base_wlan_hal_dwpal constructor and set there, not in the derived class's constructor.

tomereli replied:
The point is that the `refresh_radio_info` is implemented in the base, so I thought that in order to use @mariomaz's #977 to read the TxPower directly from nl80211 I will need to do that from `refresh_radio_info`.
It turns out that the `TxPower` we get from hostapd is fine, but nevertheless we will probably update `refresh_radio_info` to complement the radio info parameters we get from hostapd with the values we read from nl80211, so I decided to keep it there.

tomereli replied:
>More importantly though: if you move it here, it should also be taken as an argument in the base_wlan_hal_dwpal constructor and set there, not in the derived class's constructor.

What's wrong with instantiating it in `base_wlan_hal_dwpal` like I did here? It is not set in the derived class constructor:
```
base_wlan_hal_dwpal::base_wlan_hal_dwpal(HALType type, std::string iface_name,
                                         hal_event_cb_t callback, hal_conf_t hal_conf)
    : base_wlan_hal(type, iface_name, IfaceType::Intel, callback, hal_conf),
      beerocks::beerocks_fsm<dwpal_fsm_state, dwpal_fsm_event>(dwpal_fsm_state::Delay),
      m_nl80211_client(nl80211_client_factory::create_instance())
{
    LOG_IF(!m_nl80211_client, FATAL) << "Failed to create nl80211_client instance";
    // Initialize the FSM
    fsm_setup();
}
```

arnout replied:
I'm sorry, I misread the code.

arnout reviewed (approved):
Still looks good to me, with one minor comment.
